update for onedal df documentation (#1365)
oneapi-src/oneDAL: docs/source/onedal/algorithms/ensembles/decision-forest.rst

- For each :math:`j \in J_t`, find the best split :math:`s_{j,t}` that
  partitions subset :math:`D_t` and maximizes impurity decrease
  :math:`\Delta i(t)`.
- A node is a split if this split induces a decrease of the impurity greater than or equal to the predefined value
  and the number of split nodes is less or equal to :math:`\mathrm{maxLeafNodes} – 1`.
  Get the best split :math:`s_t` that maximizes impurity decrease
  :math:`\Delta i` in all :math:`s_{j,t}` splits.
- Put a node into a sorted array, where sort criterion is the improvement in impurity :math:`\Delta i(t)|D_t|`.
  The node with maximal improvement is the first in the array. For a leaf node, the improvement in impurity is zero.
- Apply this procedure to :math:`t_L` and :math:`t_R` and grow a tree one by one getting the first element from the array
  until the array is empty.

.. _df_i_math:

Inference
---------

Given decision forest classification or regression model and vectors :math:`x_1, \ldots, x_r`,
the problem is to calculate the responses for those vectors. 

.. _df_i_math_dense_hist:

Inference methods: *Dense* and *Hist*
-------------------------------------

> *Dense* and *hist* inference methods perform prediction in the same way:
> to solve the problem for each given query vector :math:`x_i`, the

Reviewer: ```suggestion
*Dense* and *hist* inference methods perform prediction in the same way.
To solve the problem for each given query vector :math:`x_i`, the
```

algorithm does the following:

.. tabs::

  .. group-tab:: Classification

>     For each tree in the forest it finds the leaf node that gives the label. The
>     forest chooses the label :math:`y` taking the majority of trees in the forest voting for that label.

Reviewer: ```suggestion
    For each tree in the forest, it finds the leaf node that gives :math:`x_i` its label. The label :math:`y` that the majority of
    trees in the forest vote for is chosen as the predicted label for the query vector :math:`x_i`.
```

  .. group-tab:: Regression

>     For each tree in the forest it finds the leaf node that gives the response as the mean of
>     dependent variables. The forest predicts the response as the mean of responses from trees.

Reviewer: ```suggestion
    For each tree in the forest, it finds the leaf node that gives :math:`x_i` the response as the mean of
    dependent variables. The mean of responses from all trees in the forest is the predicted response for the query vector :math:`x_i`.
```

Given :math:`n` feature vectors :math:`X=\{x_1=(x_{11},\ldots,x_{1p}),\ldots,x_n=(x_{n1},\ldots,x_{np})\}` of
size :math:`p`, their non-negative observation weights :math:`W=\{w_1,\ldots,w_n\}` and :math:`n` responses :math:`Y=\{y_1,\ldots,y_n\}`, 

.. tabs::

>   .. tab:: Classification

Reviewer: ```suggestion
  .. group-tab:: Classification
```

    - :math:`y_i \in \{0, \ldots, C-1\}`, where :math:`C` is the number of classes

>   .. tab:: Regression

Reviewer: ```suggestion
  .. group-tab:: Regression
```

>   .. tab:: Regression

Reviewer: ```suggestion
  .. group-tab:: Regression
```

the problem is to build a decision forest classification or regression model.

The library uses the following algorithmic framework for the training
stage. Let :math:`S = (X, Y)` be the set of observations. Given positive
integer parameters, such as the number of trees :math:`B`, the bootstrap
parameter :math:`N = f*n`, where :math:`f` is a fraction of observations used for

> a training of one tree, and the number of features per node :math:`m`, the

Reviewer 2: 'one tree' -> 'individual tree' or 'each tree in the forest' ?

Author: fixed

- Choose randomly without replacement :math:`m` feature indices :math:`J_t \in \{0, 1, \ldots, p-1\}`.
- For each :math:`j \in J_t`, find the best split :math:`s_{j,t}` that
  partitions subset :math:`D_t` and maximizes impurity decrease
  :math:`\Delta i(t)`.
- A node is a split if this split induces a decrease of the impurity greater than or equal to the predefined value
  and the number of split nodes is less or equal to :math:`\mathrm{maxLeafNodes} – 1`.
  Get the best split :math:`s_t` that maximizes impurity decrease
  :math:`\Delta i` in all :math:`s_{j,t}` splits.
- Put a node into a sorted array, where sort criterion is the improvement in impurity :math:`\Delta i(t)|D_t|`.
  The node with maximal improvement is the first in the array. For a leaf node, the improvement in impurity is zero.
- Apply this procedure to :math:`t_L` and :math:`t_R` and grow a tree one by one getting the first element from the array
  until the array is empty.

.. _df_i_math:

Inference
---------

Given decision forest classification or regression model and vectors :math:`x_1, \ldots, x_r`,
the problem is to calculate the responses for those vectors. 

.. _df_i_math_dense:

> .. _df_i_math_hist:

Reviewer 2: It isn't user-friendly approach and bad user experience

Author: fixed

the problem is to build a decision forest classification or regression model.

> Library uses the following algorithmic framework for the training

Reviewer: ```suggestion
The library uses the following algorithmic framework for the training
```

- Choose randomly without replacement :math:`m` feature indices :math:`J_t \in \{0, 1, \ldots, p-1\}`.
- For each :math:`j \in J_t`, find the best split :math:`s_{j,t}` that
  partitions subset :math:`D_t` and maximizes impurity decrease
  :math:`\Delta i(t)`.
- A node is a split if this split induces a decrease of the impurity greater than or equal to the predefined value
  and the number of split nodes is less or equal to :math:`\mathrm{maxLeafNodes} – 1`.
  Get the best split :math:`s_t` that maximizes impurity decrease
  :math:`\Delta i` in all :math:`s_{j,t}` splits.
- Put a node into a sorted array, where sort criterion is the improvement in impurity :math:`\Delta i(t)|D_t|`.
  The node with maximal improvement is the first in the array. For a leaf node, the improvement in impurity is zero.
- Apply this procedure to :math:`t_L` and :math:`t_R` and grow a tree one by one getting the first element from the array
  until the array is empty.

.. _df_i_math:

Inference
---------

Given decision forest classification or regression model and vectors :math:`x_1, \ldots, x_r`,
the problem is to calculate the responses for those vectors. 

.. _df_i_math_dense:

> .. _df_i_math_hist:

Author: agree, but it correlates to current program design.

> .. _df_i_math_hist:

Reviewer: ok let's leave it for now but I'm think we'd better discuss this with others later

Inference methods: *Dense* and *Hist*
-------------------------------------

*Dense* and *hist* inference methods performs prediction by the same way:
to solve the problem for each given query vector :math:`x_i`, the
algorithm finds the leaf node in a tree in the forest that gives:

- the label by that tree, for classification. The
  forest chooses the label :math:`y` taking the majority of trees in the
  forest voting for that label.
- the response by that tree as the mean of
  dependent variables, for regression. The forest predicts the response as the mean

>   of responses from trees.

Author: rewrote it, please check.

:math:`i(t)`. It generally reflects the homogeneity of responses within
the subset :math:`D_t` in the node :math:`t`.

.. tabs::

>   .. tab:: Classification

Reviewer: ```suggestion
  .. group-tab:: Classification
```

   
             p_i = \frac{\sum_{d \in \{d \in D | y_d = i \}} W_d}{\sum_{d \in D} W_d}
   

>   .. tab:: Regression

Reviewer: ```suggestion
  .. group-tab:: Regression
```

to solve the problem for each given query vector :math:`x_i`, the
algorithm does the following:

.. tabs::

>   .. tab:: Classification

Reviewer: ```suggestion
  .. group-tab:: Classification
```

    For each tree in the forest it finds the leaf node that gives the label. The
    forest chooses the label :math:`y` taking the majority of trees in the forest voting for that label.

>   .. tab:: Regression

Reviewer: ```suggestion
  .. group-tab:: Regression
```

The estimate of the generalization error based on the training
data can be obtained and calculated as follows:

.. tabs::

>   .. tab:: Classification

Reviewer: ```suggestion
  .. group-tab:: Classification
```

    - If OOB error value per each observation is required, then calculate the prediction error for :math:`x_i`: :math:`OOB(x_i) = I\{{y}_{i}\ne \hat{{y}_{i}}\}`

>   .. tab:: Regression

Reviewer: ```suggestion
  .. group-tab:: Regression
```
